### PR TITLE
add kindnetd as a networking option for capd

### PR DESCRIPTION
**Pull request by jaxesn**
*Issue #, if available:*

*Description of changes:*
For testing capd on arm, we need to use kindnetd since we do not currently have cilium arm builds (coming later this year).  We have also talked about making kindnetd the default for capd clusters to avoid any weirdness due to bpf/cilium.  This adds support for kindnet to the cli and the release bundle changes necessary to include the manifest.

/hold
still working on release changes

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/cluster/spec.go [hunk 1]
```diff
@@ -589,6 +589,9 @@ func (vb *VersionsBundle) Manifests() map[string][]v1alpha1.Manifest {
 	// Cilium manifest
 	manifests["cilium"] = []v1alpha1.Manifest{vb.Cilium.Manifest}
 
+	// Kindnetd manifest
+	manifests["kind"] = []v1alpha1.Manifest{vb.Kindnetd.Manifest}
```
Comment (jaxesn) [inline pkg/cluster/spec.go:593]: should this be kind or kindnetd?

File: release/pkg/assets_kindnetd.go [hunk 2]
```diff
@@ -50,5 +51,77 @@ func (r *ReleaseConfig) GetKindnetdAssets() ([]Artifact, error) {
...
+			bundleManifestArtifact := anywherev1alpha1.Manifest{
+				URI: manifestArtifact.ReleaseCdnURI,
+			}
+
+			bundleManifestArtifacts[manifestArtifact.ReleaseName] = bundleManifestArtifact
+		}
+	}
+
+	bundle := anywherev1alpha1.KindnetdBundle{
+		// Version:  ciliumGitTag,
```
Comment (jaxesn) [inline release/pkg/assets_kindnetd.go:122]: what should the version be? does this relate to the work @taneyland  is doing?

File: pkg/cluster/spec.go (hunk 1 above)
Comment (g-gaston) [reply pkg/cluster/spec.go:593]: Since the "bundle" is called `Kindnetd` I would go with kindnetd

File: pkg/networking/kindnetd.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+			return nil, fmt.Errorf("unmarhsaling kindnetd type [%s]: %v", template, err)
+		}
+		if u.GetKind() == "DaemonSet" {
+			updated, err := updatePodSubnet(clusterSpec, u)
+			if err != nil {
+				return nil, fmt.Errorf("updating kindnetd pod subnet [%s]: %v", template, err)
+			}
+			template = "\n" + strings.TrimSuffix(string(updated), "\n")
+		} else {
+			template += "---"
```
Comment (g-gaston) [inline pkg/networking/kindnetd.go:40]: nit: I think you could simplify the logic by using the helper `templater.AppendYamlResources`

File: pkg/networking/kindnetd.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	}
+	return []byte(finalContent), nil
+}
+
+func updatePodSubnet(clusterSpec *cluster.Spec, unstructured *unstructured.Unstructured) ([]byte, error) {
+	var daemonSet appsv1.DaemonSet
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructured.UnstructuredContent(), &daemonSet); err != nil {
+		return nil, fmt.Errorf("unmarhsaling kindnetd daemonset: %v", err)
+	}
+	for idx, env := range daemonSet.Spec.Template.Spec.Containers[0].Env {
```
Comment (g-gaston) [inline pkg/networking/kindnetd.go:52]: nit: maybe just check the length is `> 0`?. just in case

**Review by g-gaston — APPROVED** [review]

File: pkg/networking/kindnetd.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/yaml"
+
+	"github.com/aws/eks-anywhere/pkg/cluster"
+)
+
+type Kindnetd struct{}
```
Comment (g-gaston) [inline pkg/networking/kindnetd.go:15]: curious why kindnetd instead of kindnet?

Comment (jaxesn) [reply pkg/networking/kindnetd.go:15]: thats what they call it upstream?

File: pkg/dependencies/factory.go [hunk 6]
```diff
@@ -274,13 +274,16 @@ func (f *Factory) WithTroubleshoot() *Factory {
...
 
-func (f *Factory) WithNetworking() *Factory {
+func (f *Factory) WithNetworking(clusterConfig *v1alpha1.Cluster) *Factory {
 	f.buildSteps = append(f.buildSteps, func() error {
 		if f.dependencies.Networking != nil {
 			return nil
 		}
-
-		f.dependencies.Networking = networking.NewCilium()
+		if clusterConfig.Spec.ClusterNetwork.CNI == "kindnetd" {
```
Comment (jaxesn) [inline pkg/dependencies/factory.go:282]: is there a way to use the [constant](https://github.com/aws/eks-anywhere/pull/702/files#diff-e930b9692e6f9562202b02e50f1e6c4690205ca60ae79dd9dbb96c849e716459R304)

Comment (g-gaston) [reply pkg/dependencies/factory.go:282]: `v1alpha1.Kindnetd`

**Review by g-gaston — APPROVED** [review]

File: pkg/networking/kindnetd.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	content, err := loadManifest(clusterSpec, clusterSpec.VersionsBundle.Kindnetd.Manifest)
+	if err != nil {
+		return nil, fmt.Errorf("can't load kindnetd manifest: %v", err)
+	}
+	templates := strings.Split(string(content), "---")
+	finalTemplates := make([][]byte, 0, len(templates))
+	for _, template := range templates {
+		u := &unstructured.Unstructured{}
+		if err := yaml.Unmarshal([]byte(template), u); err != nil {
+			return nil, fmt.Errorf("unmarhsaling kindnetd type [%s]: %v", template, err)
```
Comment (abhay-krishna) [inline pkg/networking/kindnetd.go:33]:
```suggestion
			return nil, fmt.Errorf("unmarshaling kindnetd type [%s]: %v", template, err)
```

File: pkg/networking/kindnetd.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+			finalTemplates = append(finalTemplates, []byte(template))
+		}
+	}
+	return templater.AppendYamlResources(finalTemplates...), nil
+}
+
+func updatePodSubnet(clusterSpec *cluster.Spec, unstructured *unstructured.Unstructured) ([]byte, error) {
+	var daemonSet appsv1.DaemonSet
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructured.UnstructuredContent(), &daemonSet); err != nil {
+		return nil, fmt.Errorf("unmarhsaling kindnetd daemonset: %v", err)
```
Comment (abhay-krishna) [inline pkg/networking/kindnetd.go:51]:
```suggestion
		return nil, fmt.Errorf("unmarshaling kindnetd daemonset: %v", err)
```

**Review by abhay-krishna — APPROVED** [review]
/lgtm

**Pull request merged**